### PR TITLE
Add regular index-based sampler - part 1

### PR DESCRIPTION
**Pull request by NicolasHug**
This PR adds an index-based sampler that returns equally-spaced clips.
It is a partial implementation, just like the random sampler in https://github.com/pytorch/torchcodec/pull/221. In particular it is missing 2 key parts:

- The implementation of a policy to deal with edge cases when the number of available frames is smaller than what is requested by the sampler
- fast and efficient "sort and decode indices" strategy

Both these parts are left as TODO and are tracked in https://github.com/pytorch/torchcodec/issues/239. I am hoping to merge this PR *before* we tackle those, because I want to make sure the solution we come up with will satisfy both samplers (and not just the existing random sampler).

### CODE REVIEW

File: src/torchcodec/samplers/_implem.py [hunk 1]
```diff
@@ -140,3 +140,61 @@ def clips_at_random_indices(
     random.setstate(builtin_random_state)
 
     return clips
+
+
+def clips_at_regular_indices(
+    decoder: VideoDecoder,
+    *,
+    num_clips: int = 1,
```
Comment (NicolasHug) [inline src/torchcodec/samplers/_implem.py:148]: I feel like `clips_at_regular_indices(num_clips=1)` is an anti pattern. Same with 2. Maybe we should make 3 the default, and enforce `num_clips>=3`?

I don't feel too strongly about it either way - just being pedantic.

Comment (ahmadsharif1) [reply src/torchcodec/samplers/_implem.py:148]: If this is what other libraries do we can do the same here.

I was thinking of sampling the middle of the video if num_clips=1 but that would make it slower than existing libraries

File: test/samplers/test_samplers.py [hunk 2]
```diff
@@ -150,59 +165,87 @@ def test_random_sampler_randomness():
     assert builtin_random_state_start == builtin_random_state_end
 
 
-def test_random_sampler_errors():
+@pytest.mark.parametrize(
+    "num_clips, sampling_range_size",
+    (
+        # Ask for 50 clips while the sampling range is 10 frames wide
```
Comment (ahmadsharif1) [inline test/samplers/test_samplers.py:171]: Is this 50 or 10?

File: test/samplers/test_samplers.py [hunk 3]
```diff
@@ -150,59 +165,87 @@ def test_random_sampler_randomness():
...
+        sampling_range_end=sampling_range_size,  # because sampling_range_start=0
+    )
+
+    assert len(clips) == num_clips
+
+    clip_starts_seconds = torch.tensor([clip.pts_seconds[0] for clip in clips])
+    assert len(torch.unique(clip_starts_seconds)) == sampling_range_size
+
+    # Assert clips starts are ordered, i.e. the start indices don't just "wrap
+    # around". They're duplicated *and* ordered.
```
Comment (ahmadsharif1) [inline test/samplers/test_samplers.py:195]: Mention reason for not wrapping around?

**Review by ahmadsharif1 — APPROVED** [review]

File: src/torchcodec/samplers/_implem.py (hunk 1 above)
Comment (NicolasHug) [reply src/torchcodec/samplers/_implem.py:148]: OK. Other libs allow `num_clips=1`. Let's revisit if needed

File: test/samplers/test_samplers.py (hunk 3 above)
Comment (NicolasHug) [reply test/samplers/test_samplers.py:195]: There's no particular reason honestly, it's just an assertion of the behavior of `torch.linspace`. I'll add a comment near the Note to clarify that this is somewhat arbitrary.

**Pull request merged**